### DISCUSSION TITLE
AMD/ROCm support

### DISCUSSION THREAD

**Discussion by noughtmare** [hlky/stable-diffusion-webui#371] [General]
This is both a request for better integration and a report that it mostly on my RX 5600xt (6 GB VRAM) on Linux.

Two special steps are necessary:

1. In the conda environment install pytorch for rocm using pip with the command from [the pytorch.org website](https://pytorch.org/get-started/locally/), currently:
      ```
      pip3 install --upgrade torch torchvision torchaudio --extra-index-url https://download.pytorch.org/whl/rocm5.1.1
      ```
      (Note: the `--upgrade` part must be added manually)
2. Use `HSA_OVERRIDE_GFX_VERSION=10.3.0 python scripts/webui.py` in relauncher.py. This is not required for AMD GPUs that have official rocm support.

This also required me to use optimized mode and `--precision full --no-half` as mentioned [on this wiki page](https://github.com/hlky/stable-diffusion-webui/wiki/Optimized-mode#some-users-need-to-run-with---precision-full---no-half-it-is-reported-commonly-by-16xx-nvidia-graphics-card-users).

One thing that doesn't work is the CUDA memory measurement, but that is not a deal breaker. Perhaps something like rocm-smi can be used to measure memory usage.

**Comment by nderscore** [comment]
Instead of `--upgrade` it's probably better to use `--force-reinstall` maybe?

The integration works fine with ROCm though outside of memory measurement being skipped, which is a pretty low-priority feature (and it might even be a small performance gain to skip that code anyway :zipper_mouth_face: )

**Comment by heistak** [comment]
This should be able to be integrated by having a choice to use an alternative environment.yaml file for conda to use, with the following differences.

Around line 8, comment out the following lines:
```
#  - cudatoolkit=11.3
#  - pytorch=1.11.0
#  - torchvision=0.12.0

```
Around line 13, right under the beginning of the "pip:" subsection, insert the following lines:
```
    - --extra-index-url https://download.pytorch.org/whl/rocm5.1.1
    - torch # will be a version for ROCm
    - torchvision # will be a version for ROCm
    - torchaudio # will be a version for ROCm
```

I was experimenting with the installation process a few days ago and wrote a [version of environment.yaml](https://github.com/heistak/stable-diffusion-radeon/blob/main/environment2.yaml) which makes these changes (This one doesn't contain the GUI-related packages like gradio though). I presume the installer could ask the user which GPU type to use, and use the corresponding environment.yaml file.

**Reply by ponut64** [reply]
How can I get it to stick? Every time I write in these changes and re-get the dependencies, it resets the environment.yaml to the original. Where is it getting the original?!

/e: Okay, it's git stuff. Obviously, you use ``git add environment.yaml`` and then ``git stash pop``, then delete the junk in the file to resolve the discrepancy, then when you delete the ``z_version_env.tmp`` file and run webui.cmd again, it should inform you that you have unmerged files. At that point it won't change what you did in environment.yaml and should grab the specified dependencies. Will it work? Don't know!

/e2: It did not work because the environment.yaml in /src/taming-transofmers/ is still including CUDA and I don't know how to make it stop doing that.

/e3: taming-transformers is another repository. If you link it to a fork instead of that repo, make the required changes to the fork, then apply that line to pip, perhaps you will get what you want...

/e4: It does not work. Someone has to do a lot of work to fix this.

**Comment by 333377777** [comment]
Hello Me too 5600xt， cant work on ubuntu ，GPU100% use
The Code has no error feedback，So use 5600xt, must change to Linux（Excluding Ubuntu）？
